File: src/mcp/contract.py
"""
MCP Gateway Contract Normalization

Payload normalization, quality defaulting, and detection evasion guards
for the MCP gateway.
"""

import math
import os
from typing import Any, Dict, Optional

from src.config import get_config_value
from src.workflow.levels.types import ANALYSIS_SCHEMA_VERSION, LEGACY_DECISION_MAP, ensure_contract_payload
from starlette.responses import JSONResponse


# ============ Contract Wrappers ============

def _with_contract(payload: Dict[str, Any]) -> Dict[str, Any]:
    """Ensure payload has contract fields."""
    return ensure_contract_payload(payload)


def _with_terminal_contract(payload: Dict[str, Any]) -> Dict[str, Any]:
    """Ensure payload has terminal contract fields."""
    normalized = _with_contract(payload)
    if "decision" not in normalized:
        normalized["decision"] = "go"
    if "terminal" not in normalized:
        normalized["terminal"] = "done"

    terminal = normalized.get("terminal")
    legacy_terminal = terminal
    if terminal == "interrupted":
        legacy_terminal = "aborted"
    elif terminal == "recovered":
        legacy_terminal = "done"

    legacy_fields = normalized.get("legacy_contract_fields")
    if not isinstance(legacy_fields, dict):
        legacy_fields = {}
    legacy_fields.setdefault("decision", normalized.get("decision"))
    legacy_fields.setdefault("terminal", legacy_terminal)
    legacy_fields.setdefault("terminal_state", legacy_terminal)
    normalized["legacy_contract_fields"] = legacy_fields
    normalized.setdefault("terminal_state", legacy_terminal)
    return normalized


# ============ Detection Evasion Guard ============

_PROHIBITED_DETECTION_KEYS = {
    "anti_detection",
    "bypass_detector",
    "pass_gptzero",
    "detector_bypass",
    "humanize_for_detector",
}

_PROHIBITED_DETECTION_TERMS = (
    "ai detection",
    "ai detector",
    "bypass detector",
    "bypass ai detector",
    "evade detector",
    "avoid detection",
    "pass gptzero",
    "检测对抗",
    "反检测",
    "规避检测",
)


def _contains_detection_evasion_intent(value: Any) -> bool:
    """Recursively check for detection evasion intent in payload."""
    if isinstance(value, dict):
        for key, nested in value.items():
            key_str = str(key).lower()
            if key_str in _PROHIBITED_DETECTION_KEYS:
                return True
            if _contains_detection_evasion_intent(nested):
                return True
        return False

    if isinstance(value, list):
        return any(_contains_detection_evasion_intent(item) for item in value)

    if isinstance(value, str):
        lowered = value.lower()
        return any(term in lowered for term in _PROHIBITED_DETECTION_TERMS)

    return False


def _resolve_detection_evasion_guard_enabled() -> bool:
    """Check if detection evasion guard is enabled."""
    raw = os.getenv("NIKO_DETECTION_EVASION_GUARD")
    if raw is not None:
        return str(raw).strip().lower() in {"1", "true", "yes", "on"}
    return bool(get_config_value("gateway.detection_evasion_guard", True))


def _guard_detection_evasion_payload(
    payload: Dict[str, Any],
    enabled_override: Optional[bool] = None,
) -> Optional[JSONResponse]:
    """Guard against detection evasion requests."""
    guard_enabled = _resolve_detection_evasion_guard_enabled() if enabled_override is None else bool(enabled_override)
    if not guard_enabled:
        return None
    if _contains_detection_evasion_intent(payload):
        return JSONResponse(
            {
                "error": "DETECTION_EVASION_BLOCKED",
                "code": "COMPLIANCE_DETECTION_EVASION_BLOCKED",
                "message": "检测规避相关请求已被拦截。请改用质量增强目标（自然表达、可读性、风格一致性、逻辑连贯与可执行编辑建议）。",
            },
            status_code=400,
        )
    return None


# ============ Quality Payload Defaults ============

def _quality_default_payload() -> Dict[str, Any]:
    """Get default quality payload structure."""
    return {
        "analysis_schema_version": ANALYSIS_SCHEMA_VERSION,
        "quality_score": 0.0,
        "issues": [],
        "metrics": {
            "dialogue_ratio": 0.0,
            "conflict_points": 0,
            "visual_details": 0,
            "template_sentence_ratio": 0.0,
            "dimension_scores": {
                "repetition": 0.0,
                "tone": 0.0,
                "clarity": 0.0,
                "causality": 0.0,
                "detail": 0.0,
                "factuality": 0.0,
            },
            "retrieval": {
                "stage1_candidates": 0,
                "stage2_selected": 0,
                "cited_count": 0,
                "effective_hit_rate": 0.0,
            },
            "context_budget": {
                "token_total": 0,
                "token_effective": 0,
                "utilization": 0.0,
            },
            "self_learning": {
                "strategy_adoption_rate": 0.0,
                "reflector_triggered": False,
                "curator_applied": False,
            },
        },
        "publish_recommendation": "revise",
    }


# ============ Normalization Helpers ============

def _clamp_float(value: float, low: float, high: float) -> float:
    """Clamp float value to range."""
    return max(low, min(high, value))


def _safe_float(value: Any, default: float) -> float:
    """Safely convert value to float."""
    if isinstance(value, bool):
        return float(default)
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        return float(default)
    if not math.isfinite(parsed):
        return float(default)
    return parsed


def _safe_int(value: Any, default: int) -> int:
    """Safely convert value to int."""
    if isinstance(value, bool):
        return int(default)
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        return int(default)
    return parsed


def _normalize_quality_score(value: Any, default: float) -> float:
    """Normalize quality score to 0-100 range."""
    return _clamp_float(_safe_float(value, default), 0.0, 100.0)


def _normalize_ratio(value: Any, default: float) -> float:
    """Normalize ratio to 0-1 range."""
    return _clamp_float(_safe_float(value, default), 0.0, 1.0)


def _normalize_count(value: Any, default: int) -> int:
    """Normalize count to non-negative integer."""
    return max(0, _safe_int(value, default))


def _normalize_issue_text(value: Any, default: str) -> str:
    """Normalize issue text field."""
    if value is None:
        return default
    normalized = str(value).strip()
    return normalized if normalized else default


def _normalize_issue_severity(value: Any) -> str:
    """Normalize issue severity to valid values."""
    normalized = _normalize_issue_text(value, "medium").lower()
    if normalized in {"low", "medium", "high"}:
        return normalized
    return "medium"


def _normalize_issue_item(issue: Dict[str, Any]) -> Dict[str, str]:
    """Normalize issue item structure."""
    return {
        "severity": _normalize_issue_severity(issue.get("severity")),
        "type": _normalize_issue_text(issue.get("type"), "unknown"),
        "evidence": _normalize_issue_text(issue.get("evidence"), ""),
        "suggestion": _normalize_issue_text(issue.get("suggestion"), ""),
    }


def _normalize_retrieval_metrics(value: Any, fallback: Dict[str, Any]) -> Dict[str, Any]:
    """Normalize retrieval metrics structure."""
    if not isinstance(value, dict):
        value = {}
    return {
        "stage1_candidates": _normalize_count(value.get("stage1_candidates"), fallback.get("stage1_candidates", 0)),
        "stage2_selected": _normalize_count(value.get("stage2_selected"), fallback.get("stage2_selected", 0)),
        "cited_count": _normalize_count(value.get("cited_count"), fallback.get("cited_count", 0)),
        "effective_hit_rate": _normalize_ratio(value.get("effective_hit_rate"), fallback.get("effective_hit_rate", 0.0)),
    }


def _normalize_context_budget_metrics(value: Any, fallback: Dict[str, Any]) -> Dict[str, Any]:
    """Normalize context budget metrics structure."""
    if not isinstance(value, dict):
        value = {}
    return {
        "token_total": _normalize_count(value.get("token_total"), fallback.get("token_total", 0)),
        "token_effective": _normalize_count(value.get("token_effective"), fallback.get("token_effective", 0)),
        "utilization": _normalize_ratio(value.get("utilization"), fallback.get("utilization", 0.0)),
    }


def _normalize_self_learning_metrics(value: Any, fallback: Dict[str, Any]) -> Dict[str, Any]:
    """Normalize self-learning metrics structure."""
    if not isinstance(value, dict):
        value = {}
    return {
        "strategy_adoption_rate": _normalize_ratio(
            value.get("strategy_adoption_rate"),
            fallback.get("strategy_adoption_rate", 0.0),
        ),
        "reflector_triggered": bool(value.get("reflector_triggered", fallback.get("reflector_triggered", False))),
        "curator_applied": bool(value.get("curator_applied", fallback.get("curator_applied", False))),
    }


def _normalize_schema_version(payload: Dict[str, Any], contract_payload: Dict[str, Any]) -> str:
    """Normalize schema version from multiple sources."""
    candidates = [
        payload.get("analysis_schema_version"),
        payload.get("contract_version"),
        contract_payload.get("analysis_schema_version"),
        ANALYSIS_SCHEMA_VERSION,
    ]
    for candidate in candidates:
        if candidate is None:
            continue
        normalized = str(candidate).strip()
        if normalized:
            return normalized
    return ANALYSIS_SCHEMA_VERSION


def _normalize_publish_recommendation(payload: Dict[str, Any], fallback: str) -> str:
    """Normalize publish recommendation from decision fields."""
    decision_to_publish = {
        "go": "pass",
        "soft_go": "revise",
        "no_go": "block",
    }

    candidates = [
        payload.get("publish_recommendation"),
        payload.get("decision"),
        payload.get("decision_result"),
    ]
    for candidate in candidates:
        if not isinstance(candidate, str):
            continue
        normalized = candidate.strip().lower()
        if normalized in {"pass", "revise", "block"}:
            return normalized
        mapped_decision = LEGACY_DECISION_MAP.get(normalized)
        if mapped_decision in decision_to_publish:
            return decision_to_publish[mapped_decision]

    return fallback


def _normalize_quality_payload(payload: Any) -> Dict[str, Any]:
    """Normalize quality payload to standard structure."""
    fallback = _quality_default_payload()
    if not isinstance(payload, dict):
        payload = {}

    raw_metrics = payload.get("metrics")
    if not isinstance(raw_metrics, dict):
        raw_metrics = {}

    raw_dim_scores = raw_metrics.get("dimension_scores")
    if not isinstance(raw_dim_scores, dict):
        raw_dim_scores = {}

    fallback_metrics = fallback["metrics"]
    fallback_dim_scores = fallback_metrics["dimension_scores"]

    normalized_dim_scores = {
        key: _normalize_quality_score(raw_dim_scores.get(key), default)
        for key, default in fallback_dim_scores.items()
    }

    normalized_metrics = {
        "dialogue_ratio": _normalize_ratio(raw_metrics.get("dialogue_ratio"), fallback_metrics["dialogue_ratio"]),
        "conflict_points": _normalize_count(raw_metrics.get("conflict_points"), fallback_metrics["conflict_points"]),
        "visual_details": _normalize_count(raw_metrics.get("visual_details"), fallback_metrics["visual_details"]),
        "template_sentence_ratio": _normalize_ratio(
            raw_metrics.get("template_sentence_ratio"),
            fallback_metrics["template_sentence_ratio"],
        ),
        "dimension_scores": normalized_dim_scores,
        "retrieval": _normalize_retrieval_metrics(raw_metrics.get("retrieval"), fallback_metrics["retrieval"]),
        "context_budget": _normalize_context_budget_metrics(
            raw_metrics.get("context_budget"),
            fallback_metrics["context_budget"],
        ),
        "self_learning": _normalize_self_learning_metrics(
            raw_metrics.get("self_learning"),
            fallback_metrics["self_learning"],
        ),
    }

    raw_issues = payload.get("issues")
    if not isinstance(raw_issues, list):
        raw_issues = []
    normalized_issues = [_normalize_issue_item(item) for item in raw_issues if isinstance(item, dict)]

    contract_payload = ensure_contract_payload(payload)
    if not isinstance(contract_payload, dict):
        contract_payload = {}
    schema_version = _normalize_schema_version(payload, contract_payload)

    return {
        "analysis_schema_version": schema_version,
        "quality_score": _normalize_quality_score(payload.get("quality_score"), fallback["quality_score"]),
        "issues": normalized_issues,
        "metrics": normalized_metrics,
        "publish_recommendation": _normalize_publish_recommendation(payload, fallback["publish_recommendation"]),
    }


def _merge_quality_sidecar(
    result: Any,
    retrieval_metadata: Any,
    context_budget: Any,
    self_learning: Any = None,
) -> Dict[str, Any]:
    """Merge quality sidecar metadata into result payload."""
    payload = result if isinstance(result, dict) else {}
    merged = dict(payload)

    metrics = merged.get("metrics")
    if not isinstance(metrics, dict):
        metrics = {}
    metrics = dict(metrics)

    if isinstance(retrieval_metadata, dict):
        metrics["retrieval"] = retrieval_metadata
    if isinstance(context_budget, dict):
        metrics["context_budget"] = context_budget
    if isinstance(self_learning, dict):
        metrics["self_learning"] = self_learning

    merged["metrics"] = metrics
    return merged


__all__ = [
    # Contract wrappers
    "_with_contract",
    "_with_terminal_contract",
    # Detection evasion guard
    "_PROHIBITED_DETECTION_KEYS",
    "_PROHIBITED_DETECTION_TERMS",
    "_contains_detection_evasion_intent",
    "_resolve_detection_evasion_guard_enabled",
    "_guard_detection_evasion_payload",
    # Quality defaults
    "_quality_default_payload",
    # Normalization helpers
    "_clamp_float",
    "_safe_float",
    "_safe_int",
    "_normalize_quality_score",
    "_normalize_ratio",
    "_normalize_count",
    "_normalize_issue_text",
    "_normalize_issue_severity",
    "_normalize_issue_item",
    "_normalize_retrieval_metrics",
    "_normalize_context_budget_metrics",
    "_normalize_self_learning_metrics",
    "_normalize_schema_version",
    "_normalize_publish_recommendation",
    "_normalize_quality_payload",
    "_merge_quality_sidecar",
]

File: src-ts/mcp/contract.ts
/**
 * MCP Gateway Contract Normalization
 *
 * Payload normalization, quality defaulting, and detection evasion guards
 * for the MCP gateway.
 *
 * Migrated from src/mcp/contract.py
 */

import { getConfigValue, isTruthyString } from './config';
import {
  ANALYSIS_SCHEMA_VERSION,
  LEGACY_DECISION_MAP,
  ensureContractPayload,
} from '../workflow/types';

// ============================================================
// Canonical workflow contract definitions (re-exported from workflow/types)
// ============================================================

export { ANALYSIS_SCHEMA_VERSION, LEGACY_DECISION_MAP, ensureContractPayload };

// ============================================================
// JSONResponse replacement type
// ============================================================

export interface JsonResponse {
  statusCode: number;
  body: Record<string, unknown>;
}

// ============================================================
// Contract Wrappers
// ============================================================

/**
 * Ensure payload has contract fields.
 */
export function withContract(
  payload: Record<string, unknown>,
): Record<string, unknown> {
  return ensureContractPayload(payload);
}

/**
 * Ensure payload has terminal contract fields.
 */
export function withTerminalContract(
  payload: Record<string, unknown>,
): Record<string, unknown> {
  let normalized = withContract(payload);

  if (normalized.decision === undefined) normalized.decision = 'go';
  if (normalized.terminal === undefined) normalized.terminal = 'done';

  const terminal = normalized.terminal as string;
  let legacyTerminal = terminal;
  if (terminal === 'interrupted') legacyTerminal = 'aborted';
  else if (terminal === 'recovered') legacyTerminal = 'done';

  let legacyFields = normalized.legacy_contract_fields;
  if (typeof legacyFields !== 'object' || legacyFields === null) {
    legacyFields = {};
  }
  const lf = legacyFields as Record<string, unknown>;
  if (lf.decision === undefined) lf.decision = normalized.decision;
  if (lf.terminal === undefined) lf.terminal = legacyTerminal;
  if (lf.terminal_state === undefined) lf.terminal_state = legacyTerminal;

  normalized = { ...normalized, legacy_contract_fields: lf };
  if (normalized.terminal_state === undefined) {
    normalized.terminal_state = legacyTerminal;
  }
  return normalized;
}

// ============================================================
// Detection Evasion Guard
// ============================================================

export const PROHIBITED_DETECTION_KEYS: ReadonlySet<string> = new Set([
  'anti_detection',
  'bypass_detector',
  'pass_gptzero',
  'detector_bypass',
  'humanize_for_detector',
]);

export const PROHIBITED_DETECTION_TERMS: ReadonlyArray<string> = [
  'ai detection',
  'ai detector',
  'bypass detector',
  'bypass ai detector',
  'evade detector',
  'avoid detection',
  'pass gptzero',
  '\u68C0\u6D4B\u5BF9\u6297',
  '\u53CD\u68C0\u6D4B',
  '\u89C4\u907F\u68C0\u6D4B',
];

/**
 * Recursively check for detection evasion intent in payload.
 */
export function containsDetectionEvasionIntent(value: unknown): boolean {
  if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
    const obj = value as Record<string, unknown>;
    for (const [key, nested] of Object.entries(obj)) {
      const keyStr = key.toLowerCase();
      if (PROHIBITED_DETECTION_KEYS.has(keyStr)) return true;
      if (containsDetectionEvasionIntent(nested)) return true;
    }
    return false;
  }

  if (Array.isArray(value)) {
    return value.some((item) => containsDetectionEvasionIntent(item));
  }

  if (typeof value === 'string') {
    const lowered = value.toLowerCase();
    return PROHIBITED_DETECTION_TERMS.some((term) => lowered.includes(term));
  }

  return false;
}

/**
 * Check if detection evasion guard is enabled.
 */
export function resolveDetectionEvasionGuardEnabled(): boolean {
  const raw = process.env.NIKO_DETECTION_EVASION_GUARD;
  if (raw !== undefined) return isTruthyString(raw);
  return Boolean(getConfigValue('gateway.detection_evasion_guard', true));
}

/**
 * Guard against detection evasion requests.
 * Returns a JsonResponse if blocked, null if allowed.
 */
export function guardDetectionEvasionPayload(
  payload: Record<string, unknown>,
  enabledOverride?: boolean,
): JsonResponse | null {
  const guardEnabled =
    enabledOverride !== undefined
      ? Boolean(enabledOverride)
      : resolveDetectionEvasionGuardEnabled();

  if (!guardEnabled) return null;

  if (containsDetectionEvasionIntent(payload)) {
    return {
      statusCode: 400,
      body: {
        error: 'DETECTION_EVASION_BLOCKED',
        code: 'COMPLIANCE_DETECTION_EVASION_BLOCKED',
        message:
          '\u68C0\u6D4B\u89C4\u907F\u76F8\u5173\u8BF7\u6C42\u5DF2\u88AB\u62E6\u622A\u3002' +
          '\u8BF7\u6539\u7528\u8D28\u91CF\u589E\u5F3A\u76EE\u6807' +
          '\uFF08\u81EA\u7136\u8868\u8FBE\u3001\u53EF\u8BFB\u6027\u3001\u98CE\u683C\u4E00\u81F4\u6027\u3001' +
          '\u903B\u8F91\u8FDE\u8D2F\u4E0E\u53EF\u6267\u884C\u7F16\u8F91\u5EFA\u8BAE\uFF09\u3002',
      },
    };
  }

  return null;
}

// ============================================================
// Quality Payload Defaults
// ============================================================

/**
 * Get default quality payload structure.
 */
export function qualityDefaultPayload(): Record<string, unknown> {
  return {
    analysis_schema_version: ANALYSIS_SCHEMA_VERSION,
    quality_score: 0.0,
    issues: [],
    metrics: {
      dialogue_ratio: 0.0,
      conflict_points: 0,
      visual_details: 0,
      template_sentence_ratio: 0.0,
      dimension_scores: {
        repetition: 0.0,
        tone: 0.0,
        clarity: 0.0,
        causality: 0.0,
        detail: 0.0,
        factuality: 0.0,
      },
      retrieval: {
        stage1_candidates: 0,
        stage2_selected: 0,
        cited_count: 0,
        effective_hit_rate: 0.0,
      },
      context_budget: {
        token_total: 0,
        token_effective: 0,
        utilization: 0.0,
      },
      self_learning: {
        strategy_adoption_rate: 0.0,
        reflector_triggered: false,
        curator_applied: false,
      },
    },
    publish_recommendation: 'revise',
  };
}

// ============================================================
// Normalization Helpers
// ============================================================

/**
 * Clamp float value to range.
 */
export function clampFloat(value: number, low: number, high: number): number {
  return Math.max(low, Math.min(high, value));
}

/**
 * Safely convert value to float.
 */
export function safeFloat(value: unknown, fallback: number): number {
  if (typeof value === 'boolean') return fallback;
  const parsed = Number(value);
  if (Number.isNaN(parsed) || !Number.isFinite(parsed)) return fallback;
  return parsed;
}

/**
 * Safely convert value to int.
 */
export function safeInt(value: unknown, fallback: number): number {
  if (typeof value === 'boolean') return fallback;
  const parsed = parseInt(String(value), 10);
  if (Number.isNaN(parsed)) return fallback;
  return parsed;
}

/**
 * Normalize quality score to 0-100 range.
 */
export function normalizeQualityScore(value: unknown, fallback: number): number {
  return clampFloat(safeFloat(value, fallback), 0.0, 100.0);
}

/**
 * Normalize ratio to 0-1 range.
 */
export function normalizeRatio(value: unknown, fallback: number): number {
  return clampFloat(safeFloat(value, fallback), 0.0, 1.0);
}

/**
 * Normalize count to non-negative integer.
 */
export function normalizeCount(value: unknown, fallback: number): number {
  return Math.max(0, safeInt(value, fallback));
}

/**
 * Normalize issue text field.
 */
export function normalizeIssueText(value: unknown, fallback: string): string {
  if (value === null || value === undefined) return fallback;
  const normalized = String(value).trim();
  return normalized || fallback;
}

/**
 * Normalize issue severity to valid values.
 */
export function normalizeIssueSeverity(value: unknown): string {
  const normalized = normalizeIssueText(value, 'medium').toLowerCase();
  if (normalized === 'low' || normalized === 'medium' || normalized === 'high') return normalized;
  return 'medium';
}

/**
 * Normalize issue item structure.
 */
export function normalizeIssueItem(issue: Record<string, unknown>): Record<string, string> {
  return {
    severity: normalizeIssueSeverity(issue.severity),
    type: normalizeIssueText(issue.type, 'unknown'),
    evidence: normalizeIssueText(issue.evidence, ''),
    suggestion: normalizeIssueText(issue.suggestion, ''),
  };
}

/**
 * Normalize retrieval metrics structure.
 */
export function normalizeRetrievalMetrics(
  value: unknown,
  fallback: Record<string, unknown>,
): Record<string, unknown> {
  const v = typeof value === 'object' && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {};
  return {
    stage1_candidates: normalizeCount(v.stage1_candidates, Number(fallback.stage1_candidates ?? 0)),
    stage2_selected: normalizeCount(v.stage2_selected, Number(fallback.stage2_selected ?? 0)),
    cited_count: normalizeCount(v.cited_count, Number(fallback.cited_count ?? 0)),
    effective_hit_rate: normalizeRatio(v.effective_hit_rate, Number(fallback.effective_hit_rate ?? 0.0)),
  };
}

/**
 * Normalize context budget metrics structure.
 */
export function normalizeContextBudgetMetrics(
  value: unknown,
  fallback: Record<string, unknown>,
): Record<string, unknown> {
  const v = typeof value === 'object' && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {};
  return {
    token_total: normalizeCount(v.token_total, Number(fallback.token_total ?? 0)),
    token_effective: normalizeCount(v.token_effective, Number(fallback.token_effective ?? 0)),
    utilization: normalizeRatio(v.utilization, Number(fallback.utilization ?? 0.0)),
  };
}

/**
 * Normalize self-learning metrics structure.
 */
export function normalizeSelfLearningMetrics(
  value: unknown,
  fallback: Record<string, unknown>,
): Record<string, unknown> {
  const v = typeof value === 'object' && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {};
  return {
    strategy_adoption_rate: normalizeRatio(
      v.strategy_adoption_rate,
      Number(fallback.strategy_adoption_rate ?? 0.0),
    ),
    reflector_triggered: Boolean(v.reflector_triggered ?? fallback.reflector_triggered ?? false),
    curator_applied: Boolean(v.curator_applied ?? fallback.curator_applied ?? false),
  };
}

/**
 * Normalize schema version from multiple sources.
 */
export function normalizeSchemaVersion(
  payload: Record<string, unknown>,
  contractPayload: Record<string, unknown>,
): string {
  const candidates = [
    payload.analysis_schema_version,
    payload.contract_version,
    contractPayload.analysis_schema_version,
    ANALYSIS_SCHEMA_VERSION,
  ];

  for (const candidate of candidates) {
    if (candidate === undefined || candidate === null) continue;
    const normalized = String(candidate).trim();
    if (normalized) return normalized;
  }
  return ANALYSIS_SCHEMA_VERSION;
}

/**
 * Normalize publish recommendation from decision fields.
 */
export function normalizePublishRecommendation(
  payload: Record<string, unknown>,
  fallback: string,
): string {
  const decisionToPublish: Record<string, string> = {
    go: 'pass',
    soft_go: 'revise',
    no_go: 'block',
  };

  const candidates = [
    payload.publish_recommendation,
    payload.decision,
    payload.decision_result,
  ];

  for (const candidate of candidates) {
    if (typeof candidate !== 'string') continue;
    const normalized = candidate.trim().toLowerCase();
    if (normalized === 'pass' || normalized === 'revise' || normalized === 'block') {
      return normalized;
    }
    const mappedDecision = LEGACY_DECISION_MAP[normalized];
    if (mappedDecision !== undefined && mappedDecision in decisionToPublish) {
      return decisionToPublish[mappedDecision];
    }
  }

  return fallback;
}

/**
 * Normalize quality payload to standard structure.
 */
export function normalizeQualityPayload(
  payload: unknown,
): Record<string, unknown> {
  const fb = qualityDefaultPayload();
  const p: Record<string, unknown> =
    typeof payload === 'object' && payload !== null && !Array.isArray(payload)
      ? payload as Record<string, unknown>
      : {};

  const rawMetrics = typeof p.metrics === 'object' && p.metrics !== null && !Array.isArray(p.metrics)
    ? p.metrics as Record<string, unknown>
    : {};

  const rawDimScores = typeof rawMetrics.dimension_scores === 'object' &&
    rawMetrics.dimension_scores !== null && !Array.isArray(rawMetrics.dimension_scores)
    ? rawMetrics.dimension_scores as Record<string, unknown>
    : {};

  const fallbackMetrics = fb.metrics as Record<string, unknown>;
  const fallbackDimScores = fallbackMetrics.dimension_scores as Record<string, number>;

  const normalizedDimScores: Record<string, number> = {};
  for (const [key, defaultVal] of Object.entries(fallbackDimScores)) {
    normalizedDimScores[key] = normalizeQualityScore(rawDimScores[key], defaultVal);
  }

  const normalizedMetrics: Record<string, unknown> = {
    dialogue_ratio: normalizeRatio(rawMetrics.dialogue_ratio, Number(fallbackMetrics.dialogue_ratio)),
    conflict_points: normalizeCount(rawMetrics.conflict_points, Number(fallbackMetrics.conflict_points)),
    visual_details: normalizeCount(rawMetrics.visual_details, Number(fallbackMetrics.visual_details)),
    template_sentence_ratio: normalizeRatio(
      rawMetrics.template_sentence_ratio,
      Number(fallbackMetrics.template_sentence_ratio),
    ),
    dimension_scores: normalizedDimScores,
    retrieval: normalizeRetrievalMetrics(
      rawMetrics.retrieval,
      fallbackMetrics.retrieval as Record<string, unknown>,
    ),
    context_budget: normalizeContextBudgetMetrics(
      rawMetrics.context_budget,
      fallbackMetrics.context_budget as Record<string, unknown>,
    ),
    self_learning: normalizeSelfLearningMetrics(
      rawMetrics.self_learning,
      fallbackMetrics.self_learning as Record<string, unknown>,
    ),
  };

  const rawIssues = Array.isArray(p.issues) ? p.issues : [];
  const normalizedIssues = rawIssues
    .filter((item): item is Record<string, unknown> =>
      typeof item === 'object' && item !== null && !Array.isArray(item))
    .map(normalizeIssueItem);

  let contractPayload = ensureContractPayload(p);
  if (typeof contractPayload !== 'object' || contractPayload === null) {
    contractPayload = {};
  }
  const schemaVersion = normalizeSchemaVersion(p, contractPayload);

  return {
    analysis_schema_version: schemaVersion,
    quality_score: normalizeQualityScore(p.quality_score, Number(fb.quality_score)),
    issues: normalizedIssues,
    metrics: normalizedMetrics,
    publish_recommendation: normalizePublishRecommendation(p, String(fb.publish_recommendation)),
  };
}

/**
 * Merge quality sidecar metadata into result payload.
 */
export function mergeQualitySidecar(
  result: unknown,
  retrievalMetadata: unknown,
  contextBudget: unknown,
  selfLearning?: unknown,
): Record<string, unknown> {
  const payload: Record<string, unknown> =
    typeof result === 'object' && result !== null && !Array.isArray(result)
      ? { ...(result as Record<string, unknown>) }
      : {};

  let metrics: Record<string, unknown> =
    typeof payload.metrics === 'object' && payload.metrics !== null && !Array.isArray(payload.metrics)
      ? { ...(payload.metrics as Record<string, unknown>) }
      : {};

  if (typeof retrievalMetadata === 'object' && retrievalMetadata !== null && !Array.isArray(retrievalMetadata)) {
    metrics = { ...metrics, retrieval: retrievalMetadata };
  }
  if (typeof contextBudget === 'object' && contextBudget !== null && !Array.isArray(contextBudget)) {
    metrics = { ...metrics, context_budget: contextBudget };
  }
  if (typeof selfLearning === 'object' && selfLearning !== null && !Array.isArray(selfLearning)) {
    metrics = { ...metrics, self_learning: selfLearning };
  }

  return { ...payload, metrics };
}
